[java] Adds support for OrtEnvironment thread pools

Craigacp opened the pull request:
**Description**:
Adds support for environment level thread pools to the Java API. It also bumps a few things in the gradle file to make it support newer gradle versions (though it doesn't bump the gradle version itself).

**Motivation and Context**
- Why is this change required? What problem does it solve? Java <-> C API parity.

gwang-msft commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux CPU x64 NoContribops CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, MacOS NoContribops CI Pipeline, Windows CPU CI Pipeline

Craigacp commented:
Looks like I messed up the cuda conditional test runner. That can be fixed easily. Not sure why the cpu tests failed, it looks like it happened somewhere in android?

yuslepukhin commented:
> #set(CMAKE_VERBOSE_MAKEFILE on)

Does not seem to have changes

---
Refers to: cmake/onnxruntime_java.cmake:4 in 5ae1562. [](commit_id = 5ae1562d843ce7495f040195f7690cccbcbaf399, deletion_comment = False)

yuslepukhin commented on java/src/main/native/ai_onnxruntime_OrtEnvironment.c:
```diff
@@ -24,6 +25,27 @@ JNIEXPORT jlong JNICALL Java_ai_onnxruntime_OrtEnvironment_createHandle(JNIEnv *
+ */
+JNIEXPORT jlong JNICALL Java_ai_onnxruntime_OrtEnvironment_createHandle__JILjava_lang_String_2J
+  (JNIEnv * jniEnv, jclass jobj, jlong apiHandle, jint loggingLevel, jstring name, jlong threadOptionsHandle) {
+    (void) jobj; // Required JNI parameter not needed by functions which don't need to access their host object.
```
> jobj [](start = 10, length = 5)

What if you just commented out the param name?

Craigacp replied:
The function signature is generated by JNI, I can't control it. And the compiler errors if there are unused arguments.

Craigacp replied:
Oh wait, I see what you mean. I'll check.

yuslepukhin commented on java/src/main/native/ai_onnxruntime_OrtEnvironment.c:
```diff
@@ -12,7 +12,8 @@
  * Signature: (Ljava/lang/String;)J
  */
-JNIEXPORT jlong JNICALL Java_ai_onnxruntime_OrtEnvironment_createHandle(JNIEnv * jniEnv, jclass jobj, jlong apiHandle, jint loggingLevel, jstring name) {
+JNIEXPORT jlong JNICALL Java_ai_onnxruntime_OrtEnvironment_createHandle__JILjava_lang_String_2
```
>__JILjava_lang_String_2 [](start = 71, length = 23)

Out of scientific curiosity, why the name change?

Craigacp replied:
The names are automatically mangled by JNI, as it's emitting a C header and we don't have function overloading. There are two createHandle methods now, one with a threading options, one without. I guess I could pass a 0 in and unify them, but they call different functions in the c api so it wouldn't be much shorter.

yuslepukhin commented on java/src/test/java/ai/onnxruntime/InferenceTest.java:
```diff
@@ -393,6 +394,75 @@ public void createSessionFromByteArray() throws IOException, OrtException {
+        OrtEnvironment newEnv =
+            OrtEnvironment.getEnvironment(
+                OrtLoggingLevel.ORT_LOGGING_LEVEL_FATAL, "fail", threadOpts);
+        fail("Should have thrown IllegalStateException");
```
> fail("Should have thrown IllegalStateException"); [](start = 7, length = 50)

Needs comment as the reason for the failure

yuslepukhin commented on java/src/main/java/ai/onnxruntime/OrtEnvironment.java:
```diff
@@ -110,6 +110,36 @@ public static synchronized OrtEnvironment getEnvironment(
+   * Creates an OrtEnvironment using the specified global thread pool options. Note unlike the other
+   * {@code getEnvironment} methods if there already is an existing OrtEnvironment this call throws
+   * {@link IllegalStateException} as we cannot guarantee that the environment has the appropriate
+   * thread pool configuration.
```
>ation. [](start = 25, length = 6)

Still throws if already created.

Craigacp replied:
I'm not sure what you mean? As far as I can tell there's no way to check if an environment has a thread pool in it, nor what that pool looks like. I guess I could cache the options, mirror all the fields in Java and then do an equality test as part of this check?

yuslepukhin replied:
Disregard this comment. The behavior has changed for the better lately.

---
In reply to: [562983378](https://github.com/microsoft/onnxruntime/pull/6406#discussion_r562983378) [](ancestors = 562983378)

yuslepukhin commented:
InferenceTest > testModelSerialization() FAILED
2021-01-22T21:38:50.1614760Z     java.lang.UnsatisfiedLinkError at InferenceTest.java:1414
2021-01-22T21:38:50.1615150Z 
2021-01-22T21:38:50.1615688Z InferenceTest > testModelInputBuffer() FAILED
2021-01-22T21:38:50.1616243Z     java.lang.NoClassDefFoundError at InferenceTest.java:917
2021-01-22T21:38:50.1616631Z 
2021-01-22T21:38:50.1617085Z InferenceTest > testModelInputDOUBLE() FAILED
2021-01-22T21:38:50.1617647Z     java.lang.NoClassDefFoundError at InferenceTest.java:1187
2021-01-22T21:38:50.1618009Z 
2021-01-22T21:38:50.1618486Z InferenceTest > createSessionFromByteArray() FAILED
2021-01-22T21:38:50.1619044Z     java.lang.NoClassDefFoundError at InferenceTest.java:351
2021-01-22T21:38:50.1619418Z 
2021-01-22T21:38:50.1619895Z InferenceTest > repeatedCloseTest() FAILED
2021-01-22T21:38:50.1620437Z     java.lang.NoClassDefFoundError at InferenceTest.java:73
2021-01-22T21:38:50.1620816Z 
2021-01-22T21:38:50.1621243Z InferenceTest > testCUDA() SKIPPED
2021-01-22T21:38:50.1621532Z 
2021-01-22T21:38:50.1622035Z InferenceTest > testSymbolicDimensionAssignment() FAILED
2021-01-22T21:38:50.1622612Z     java.lang.NoClassDefFoundError at InferenceTest.java:667
2021-01-22T21:38:50.1622990Z 
2021-01-22T21:38:50.1623543Z InferenceTest > testModelSequenceOfMapIntFloat() FAILED
2021-01-22T21:38:50.1624152Z     java.lang.NoClassDefFoundError at InferenceTest.java:1279
2021-01-22T21:38:50.1624512Z 
2021-01-22T21:38:50.1624979Z InferenceTest > testExtraSessionOptions() FAILED
2021-01-22T21:38:50.1625543Z     java.lang.NoClassDefFoundError at InferenceTest.java:1012
2021-01-22T21:38:50.1625918Z 
2021-01-22T21:38:50.1626376Z InferenceTest > throwExtraInputs() FAILED
2021-01-22T21:38:50.1626918Z     java.lang.NoClassDefFoundError at InferenceTest.java:1511
2021-01-22T21:38:50.1627293Z 
2021-01-22T21:38:50.1627760Z InferenceTest > testModelInputFLOAT() FAILED
2021-01-22T21:38:50.1628308Z     java.lang.NoClassDefFoundError at InferenceTest.java:890
2021-01-22T21:38:50.1628666Z 
2021-01-22T21:38:50.1629108Z InferenceTest > testModelInputINT16() FAILED
2021-01-22T21:38:50.1629650Z     java.lang.NoClassDefFoundError at InferenceTest.java:1231
2021-01-22T21:38:50.1630022Z 
2021-01-22T21:38:50.1630500Z InferenceTest > testModelInputINT32() FAILED
2021-01-22T21:38:50.1631042Z     java.lang.NoClassDefFoundError at InferenceTest.java:1165
2021-01-22T21:38:50.1631419Z 
2021-01-22T21:38:50.1631866Z InferenceTest > testModelInputINT64() FAILED
2021-01-22T21:38:50.1632412Z     java.lang.NoClassDefFoundError at InferenceTest.java:1253
2021-01-22T21:38:50.1632768Z 
2021-01-22T21:38:50.1633215Z InferenceTest > testModelInputBOOL() FAILED
2021-01-22T21:38:50.1633758Z     java.lang.NoClassDefFoundError at InferenceTest.java:1143
2021-01-22T21:38:50.1634126Z 
2021-01-22T21:38:50.1634595Z InferenceTest > testModelInputINT8() FAILED
2021-01-22T21:38:50.1635133Z     java.lang.NoClassDefFoundError at InferenceTest.java:1209
2021-01-22T21:38:50.1635507Z 
2021-01-22T21:38:50.1635961Z InferenceTest > throwWrongInputName() FAILED
2021-01-22T21:38:50.1636502Z     java.lang.NoClassDefFoundError at InferenceTest.java:1511
2021-01-22T21:38:50.1636859Z 
2021-01-22T21:38:50.1637309Z InferenceTest > throwWrongInputType() FAILED
2021-01-22T21:38:50.1637846Z     java.lang.NoClassDefFoundError at InferenceTest.java:1511
2021-01-22T21:38:50.1638372Z 
2021-01-22T21:38:50.1638852Z InferenceTest > testModelSequenceOfMapStringFloat() FAILED
2021-01-22T21:38:50.1639419Z     java.lang.NoClassDefFoundError at InferenceTest.java:1346
2021-01-22T21:38:50.1639792Z 
2021-01-22T21:38:50.1640224Z InferenceTest > testProviders() FAILED
2021-01-22T21:38:50.1640761Z     java.lang.NoClassDefFoundError at InferenceTest.java:651
2021-01-22T21:38:50.1641114Z 
2021-01-22T21:38:50.1641569Z InferenceTest > testLoadCustomLibrary() FAILED
2021-01-22T21:38:50.1642115Z     java.lang.NoClassDefFoundError at InferenceTest.java:1099
2021-01-22T21:38:50.1642587Z 
2021-01-22T21:38:50.1643087Z InferenceTest > testAllPretrainedModels() SKIPPED
2021-01-22T21:38:50.1643397Z 
2021-01-22T21:38:50.1643836Z InferenceTest > testRunOptions() FAILED
2021-01-22T21:38:50.1644395Z     java.lang.NoClassDefFoundError at InferenceTest.java:978
2021-01-22T21:38:50.1644767Z 
2021-01-22T21:38:50.1645272Z InferenceTest > environmentThreadPoolTest() FAILED
2021-01-22T21:38:50.1645827Z     java.lang.UnsatisfiedLinkError at InferenceTest.java:405
2021-01-22T21:38:50.1646201Z 
2021-01-22T21:38:50.1646632Z InferenceTest > inferenceTest() FAILED
2021-01-22T21:38:50.1647194Z     java.lang.NoClassDefFoundError at InferenceTest.java:479
2021-01-22T21:38:50.1647549Z 
2021-01-22T21:38:50.1648026Z InferenceTest > morePartialInputsTest() FAILED
2021-01-22T21:38:50.1648608Z     java.lang.NoClassDefFoundError at InferenceTest.java:126
2021-01-22T21:38:50.1648977Z 
2021-01-22T21:38:50.1649433Z InferenceTest > testStringIdentity() FAILED
2021-01-22T21:38:50.1649990Z     java.lang.NoClassDefFoundError at InferenceTest.java:1432
2021-01-22T21:38:50.1650359Z 
2021-01-22T21:38:50.1650830Z InferenceTest > createSessionFromPath() FAILED
2021-01-22T21:38:50.1651370Z     java.lang.NoClassDefFoundError at InferenceTest.java:85
2021-01-22T21:38:50.1651718Z 
2021-01-22T21:38:50.1652164Z InferenceTest > partialInputsTest() FAILED
2021-01-22T21:38:50.1652702Z     java.lang.NoClassDefFoundError at InferenceTest.java:213
2021-01-22T21:38:50.1653077Z 
2021-01-22T21:38:50.1653728Z InferenceTest > testMultiThreads() FAILED
2021-01-22T21:38:50.1654327Z     java.lang.NoClassDefFoundError at InferenceTest.java:1511
2021-01-22T21:38:50.1654701Z

Craigacp commented:
I added a comment to the test, fixed the CUDA test, and tried to fix the android thing. I'm not sure what's going on there, from the log it might be because the JNI library failed to link in the line above the test failures, but there isn't any explanation as to why it failed to link. I didn't write any of the android tests, and I've never seen them run before so I don't know how they are configured.

yuslepukhin commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux CPU x64 NoContribops CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, MacOS NoContribops CI Pipeline, Windows CPU CI Pipeline

yuslepukhin commented:
/azp run Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, centos7_cpu, centos7_cpu (linux_centos_ci Debug), centos7_cpu (linux_centos_ci Release), orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline

gwang-msft commented:
/azp run Android CI Pipeline

gwang-msft commented:
/azp run Linux Nuphar CI Pipeline, orttraining-distributed

gwang-msft commented:
/azp run Linux Nuphar CI Pipeline

snnn commented:
/azp run Linux Nuphar CI Pipeline

snnn reviewed (approved):
Thank you!
